# 1 - README.md

- Reference: XGboot
https://github.com/dataquestio/project-walkthroughs/blob/master/bitcoin_price/prediction.ipynb  (https://www.youtube.com/watch?v=TF2Nx_ifmrU&t=2550s)


# 2 - Load Data

**WORK WITH TIME-SERIES DATA:**

- This data is time series it's sequential -> we don't use Cross-Validation or any of the model ML techniques to evaluate error. TimeSeriesSplit, which is a specific type of cross-validation technique used for time series data. It's important to use time-series cross-validation when dealing with sequential data to avoid training on future data. <br> TimeSeriesSplit splits the data into folds, so that the folds with data from the previous past will be used as the training set, and the future data will only be used as the test set. For example, if we split the data into 3 folds, each fold would consist of:
    - Fold 1: Data from January 2016 to December 2017 (training set) and data from January 2018 to December 2018 (test set).    
    - Fold 2: Data from January 2016 to December 2018 (training set) and data from January 2019 to December 2019 (test set).    
    - Fold 3: Data from January 2016 to December 2019 (training set) and data from January 2020 to December 2020 (test set). <br>
`tscv = TimeSeriesSplit(n_splits=3)`


In [104]:
# Load Data
import pandas as pd
data = pd.read_csv("https://raw.githubusercontent.com/lavibula/ML20222.PredictionBitcoin/main/data/data.csv")
# data = pd.read_csv(r"C:\Users\Administrator\OneDrive - Hanoi University of Science and Technology\ITE10 - Data Science and AI - HUST\20222\ML\Source_Codes\ML20222.PredictionBitcoin\data\data.csv")
# print(data.info())
data = data.sort_values('Date', ascending=True).reset_index(drop=True) # drop=True bỏ 'index' col cũ. 

print(data.info())
data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2661 entries, 0 to 2660
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    2661 non-null   object 
 1   BTC_close               2661 non-null   float64
 2   BTC_open                2661 non-null   float64
 3   BTC_high                2661 non-null   float64
 4   BTC_low                 2661 non-null   float64
 5   difficulty              2661 non-null   int64  
 6   addresses_active_count  2661 non-null   int64  
 7   sum_lock_weight         2661 non-null   int64  
 8   mean_lock_size_ytes     2661 non-null   float64
 9   total_fees_usd          2661 non-null   float64
 10  mean_hash_rate          2661 non-null   float64
 11  xfer_cnt                2661 non-null   int64  
 12  mean_tx_size_usd        2661 non-null   float64
 13  ETH                     2661 non-null   float64
 14  LTC                     2661 non-null   

Date  BTC_close  BTC_open  BTC_high  BTC_low      difficulty   
0     2016-03-10      415.8     412.8     417.5    410.3    158427203767  \
1     2016-03-11      419.1     415.8     422.4    415.1    158427203767   
2     2016-03-12      410.4     419.1     420.7    407.0    158427203767   
3     2016-03-13      412.4     410.4     415.9    409.6    158427203767   
4     2016-03-14      414.3     412.4     416.1    411.2    158427203767   
...          ...        ...       ...       ...      ...             ...   
2656  2023-06-18    26339.7   26515.0   26679.3  26290.6  52350439455487   
2657  2023-06-19    26845.9   26339.7   27029.7  26295.1  52350439455487   
2658  2023-06-20    28307.7   26845.9   28393.0  26665.5  52350439455487   
2659  2023-06-21    29996.9   28307.7   30769.5  28270.5  52350439455487   
2660  2023-06-22    29890.5   29992.8   30497.8  29590.4  52350439455487   

      addresses_active_count  sum_lock_weight  mean_lock_size_ytes   
0                     445273        426654988         6.881532e+05  \
1                     434658        398582424         6.227850e+05   
2                     374730        331208848         5.750154e+05   
3                     421585        334817852         6.293569e+05   
4                     451902        437739524         7.199663e+05   
...                      ...              ...                  ...   
2656                  883864        603064705         1.985675e+06   
2657                  920552        567091224         1.758290e+06   
2658                  951926        543133484         1.677592e+06   
2659                  920555        483181576         1.611773e+06   
2660                 1000988        547153915         1.650655e+06   

      total_fees_usd  ...      ETH    LTC      XRP      DOGE  COPPER     GOLD   
0       1.731272e+04  ...     0.00   0.00  0.00000  0.000000  0.0000     0.00  \
1       1.710193e+04  ...     0.00   0.00  0.00000  0.000000  0.0000     0.00   
2       1.398444e+04  ...     0.00   0.00  0.00000  0.000000  0.0000     0.00   
3       1.460678e+04  ...     0.00   0.00  0.00000  0.000000  0.0000     0.00   
4       1.689298e+04  ...     0.00   0.00  0.00000  0.000000  0.0000     0.00   
...              ...  ...      ...    ...      ...       ...     ...      ...   
2656    7.946708e+05  ...  1720.98  77.20  0.48699  0.062107  3.8738  1969.45   
2657    8.514319e+05  ...  1737.06  77.51  0.49341  0.062429  3.8643  1964.05   
2658    1.052364e+06  ...  1791.61  80.31  0.49270  0.063108  3.8835  1947.70   
2659    1.347976e+06  ...  1889.87  85.12  0.49995  0.065644  3.9135  1944.90   
2660    1.292778e+06  ...  1872.32  85.97  0.49401  0.065567  3.8900  1923.70   

      SILVER      SPX    JP225      DJI  
0      0.000     0.00      0.0      0.0  
1      0.000     0.00      0.0      0.0  
2      0.000     0.00      0.0      0.0  
3      0.000     0.00      0.0      0.0  
4      0.000     0.00      0.0      0.0  
...      ...      ...      ...      ...  
2656  24.198     0.00      0.0      0.0  
2657  24.062     0.00      0.0      0.0  
2658  23.234  4388.71  33155.0  34356.0  
2659  22.810  4365.69  33390.0  34235.0  
2660  22.467  4381.89  33445.0  34210.0  

[2661 rows x 23 columns]

In [105]:
data["BTC_tomorrow"] = data["BTC_close"].shift(-1)
data["BTC_target"] = (data["BTC_tomorrow"] > data["BTC_close"]).astype(int)
data["BTC_target"].value_counts()  # BTC_target: BTC_direction, compare BTC_close with BTC_tomorrow.
                                    # return 1 if nay < mai. 
data.drop('BTC_tomorrow', axis=1, inplace=True)  # Thêm `inplace=True` để áp dụng thay đổi trực tiếp vào DataFrame
print(data.info())
data


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2661 entries, 0 to 2660
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    2661 non-null   object 
 1   BTC_close               2661 non-null   float64
 2   BTC_open                2661 non-null   float64
 3   BTC_high                2661 non-null   float64
 4   BTC_low                 2661 non-null   float64
 5   difficulty              2661 non-null   int64  
 6   addresses_active_count  2661 non-null   int64  
 7   sum_lock_weight         2661 non-null   int64  
 8   mean_lock_size_ytes     2661 non-null   float64
 9   total_fees_usd          2661 non-null   float64
 10  mean_hash_rate          2661 non-null   float64
 11  xfer_cnt                2661 non-null   int64  
 12  mean_tx_size_usd        2661 non-null   float64
 13  ETH                     2661 non-null   float64
 14  LTC                     2661 non-null   

Date  BTC_close  BTC_open  BTC_high  BTC_low      difficulty   
0     2016-03-10      415.8     412.8     417.5    410.3    158427203767  \
1     2016-03-11      419.1     415.8     422.4    415.1    158427203767   
2     2016-03-12      410.4     419.1     420.7    407.0    158427203767   
3     2016-03-13      412.4     410.4     415.9    409.6    158427203767   
4     2016-03-14      414.3     412.4     416.1    411.2    158427203767   
...          ...        ...       ...       ...      ...             ...   
2656  2023-06-18    26339.7   26515.0   26679.3  26290.6  52350439455487   
2657  2023-06-19    26845.9   26339.7   27029.7  26295.1  52350439455487   
2658  2023-06-20    28307.7   26845.9   28393.0  26665.5  52350439455487   
2659  2023-06-21    29996.9   28307.7   30769.5  28270.5  52350439455487   
2660  2023-06-22    29890.5   29992.8   30497.8  29590.4  52350439455487   

      addresses_active_count  sum_lock_weight  mean_lock_size_ytes   
0                     445273        426654988         6.881532e+05  \
1                     434658        398582424         6.227850e+05   
2                     374730        331208848         5.750154e+05   
3                     421585        334817852         6.293569e+05   
4                     451902        437739524         7.199663e+05   
...                      ...              ...                  ...   
2656                  883864        603064705         1.985675e+06   
2657                  920552        567091224         1.758290e+06   
2658                  951926        543133484         1.677592e+06   
2659                  920555        483181576         1.611773e+06   
2660                 1000988        547153915         1.650655e+06   

      total_fees_usd  ...    LTC      XRP      DOGE  COPPER     GOLD  SILVER   
0       1.731272e+04  ...   0.00  0.00000  0.000000  0.0000     0.00   0.000  \
1       1.710193e+04  ...   0.00  0.00000  0.000000  0.0000     0.00   0.000   
2       1.398444e+04  ...   0.00  0.00000  0.000000  0.0000     0.00   0.000   
3       1.460678e+04  ...   0.00  0.00000  0.000000  0.0000     0.00   0.000   
4       1.689298e+04  ...   0.00  0.00000  0.000000  0.0000     0.00   0.000   
...              ...  ...    ...      ...       ...     ...      ...     ...   
2656    7.946708e+05  ...  77.20  0.48699  0.062107  3.8738  1969.45  24.198   
2657    8.514319e+05  ...  77.51  0.49341  0.062429  3.8643  1964.05  24.062   
2658    1.052364e+06  ...  80.31  0.49270  0.063108  3.8835  1947.70  23.234   
2659    1.347976e+06  ...  85.12  0.49995  0.065644  3.9135  1944.90  22.810   
2660    1.292778e+06  ...  85.97  0.49401  0.065567  3.8900  1923.70  22.467   

          SPX    JP225      DJI  BTC_target  
0        0.00      0.0      0.0           1  
1        0.00      0.0      0.0           0  
2        0.00      0.0      0.0           1  
3        0.00      0.0      0.0           1  
4        0.00      0.0      0.0           1  
...       ...      ...      ...         ...  
2656     0.00      0.0      0.0           1  
2657     0.00      0.0      0.0           1  
2658  4388.71  33155.0  34356.0           1  
2659  4365.69  33390.0  34235.0           0  
2660  4381.89  33445.0  34210.0           0  

[2661 rows x 24 columns]

In [106]:
print(data.columns)

Index(['Date', 'BTC_close', 'BTC_open', 'BTC_high', 'BTC_low', 'difficulty',
       'addresses_active_count', 'sum_lock_weight', 'mean_lock_size_ytes',
       'total_fees_usd', 'mean_hash_rate', 'xfer_cnt', 'mean_tx_size_usd',
       'ETH', 'LTC', 'XRP', 'DOGE', 'COPPER', 'GOLD', 'SILVER', 'SPX', 'JP225',
       'DJI', 'BTC_target'],
      dtype='object')


# 3 - Train Model: XGBoost Classification. (Extreme Gradient Boosting)

## 3.0 Định nghĩa and split data for pre-train

- Ensemble Methods: method of combining multiple models (search gg Machine Learning picture)    
    - Stacking techniques    
    - Bagging technique: Random Forest alg.  
    - Boosting technique: XGBoost, LightGBM, CatBoost, AdaBoost. Repeat the training process with different weights.        
        - XGBoost: regularization to reduce overfitting, gradient boosting to optimize tree weights, parallelization to train independent trees and speed up the training process, tools to monitor (giám sát) the training process.        - AdaBoost is the first boosting algorithm, CatBoost focuses on feature categorical (tính phân loại) and LightGBM focuses on speed and scalability for large datasets.

In [ ]:
# Split data for pre-train model 
train = data.iloc[:-250]
test = data.iloc[-250:]

predictors = ['BTC_close', 'BTC_open', 'BTC_high', 'BTC_low', 'difficulty',
       'addresses_active_count', 'sum_lock_weight', 'mean_lock_size_ytes',
       'total_fees_usd', 'mean_hash_rate', 'xfer_cnt', 'mean_tx_size_usd',
       'ETH', 'LTC', 'XRP', 'DOGE', 'COPPER', 'GOLD', 'SILVER', 'SPX', 'JP225',
       'DJI']  # drop BTC_tomorrow and BTC_target

## 3.1 simple RandomForestClassifier, XGBClassifier Model 

In [107]:
# simple RandomForestClassifier Model 
from sklearn.ensemble import RandomForestClassifier   #Ensemble Method
from sklearn.metrics import precision_score # ValueError: continuous is not supported, nhãn 0/1

rf_model = RandomForestClassifier(random_state=1, n_estimators=100, min_samples_split=50)
                               # số cây, individual decision tree should not split its nodes unless it has at least 50 samples. 

rf_model.fit(train[predictors], train["BTC_target"])

y_pred_test = rf_model.predict(test[predictors])
precision_score(test["BTC_target"], y_pred_test)

0.48214285714285715

In [108]:
# SIMPLE XGBClassifier Model when hyperparameters have not been adjusted 
from xgboost import XGBClassifier

xgb_model = XGBClassifier(random_state=1, learning_rate=.1, n_estimators=150)
xgb_model.fit(train[predictors], train["BTC_target"])

y_pred_test = xgb_model.predict(test[predictors])
precision_score(test["BTC_target"], y_pred_test)

0.4823529411764706

Hyperparameter Tuning

In [110]:
# Some parmaters XGB with GridSearchCV
from sklearn.model_selection import GridSearchCV, TimeSeriesSplit
from xgboost import XGBClassifier
import numpy as np

xgb = XGBClassifier() # XGBoost

# Lựa chọn giá trị siêu tham số: dựa vào data và value thông thường XGBoost
param_grid = {'random_state': [1], 
    'learning_rate': [0.1, 0.3, 0.5],
    'n_estimators': [50, 100, 150, 200], 
    'max_depth': [3,5,7], 
#      'colsample_bytree': [0.7, 0.8, 0.9],
#       'subsample': [0.7, 0.8, 0.9],
#       'gamma'
#       'min_child_samples': [1, 5, 10]
}

# thay đổi cv sang TimeSeriesSplit-specific tyoe of cv tech used for time series- 
tscv = TimeSeriesSplit(n_splits=3)
grid_search_model = GridSearchCV(estimator=xgb, # model ước lượng
                           param_grid=param_grid,
                           cv=tscv, ) # sử dụng TimeSeriesSplit

# fit, Find best hyperparameters.
grid_search_model.fit(train[predictors], train["BTC_target"])
print("Best parameters found: ", grid_search_model.best_params_)

best_xgb_model = grid_search_model.best_estimator_  #set: best_xgb_model with best_estimator. 

y_pred_test = best_xgb_model.predict(test[predictors])
precision_score(test["BTC_target"], y_pred_test)

# 0.5299145299145299

Best parameters found:  {'learning_rate': 0.5, 'max_depth': 7, 'n_estimators': 50, 'random_state': 1}


0.5299145299145299

In [111]:
# All parmaters XGB with RandomizedSearchCV
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, TimeSeriesSplit
from xgboost import XGBClassifier
import numpy as np

xgb = XGBClassifier() # XGBoost

# Lựa chọn giá trị siêu tham số: dựa vào data và value thông thường XGBoost
from scipy.stats import uniform

param_xgb = {   
    'random_state': [1],
    'learning_rate': [0.01, 0.1, 0.2, 0.3],
    'n_estimators': [i for i in range(50, 1000, 50)], # số cây, tăng độ chính xác, giảm overfitting
    'max_depth': [i for i in range(3,12)], # độ sâu tối đa
    'colsample_bytree': [0.5, 0.7, 1], # tỷ lệ features được dùng cho mỗi cây
    'subsample': [0.5, 0.7, 1], #tỷ lệ các mẫu cho mỗi cây
    'gamma': uniform(loc=0, scale=5), # cắt tỉa: dùng uniform distribution với loc = 0 và scale = 5
    'min_child_samples': [5, 10, 20] # só lượng mẫu tối thiểu được yêu cầu để tách 1 node
}


# thay đổi cv sang TimeSeriesSplit-specific tyoe of cv tech used for time series- 
tscv = TimeSeriesSplit(n_splits=3)
grid_search_model = RandomizedSearchCV(estimator=xgb, # model ước lượng
                           param_distributions=param_xgb,
                           cv=tscv, ) # sử dụng TimeSeriesSplit

# fit, Find best hyperparameters.
grid_search_model.fit(train[predictors], train["BTC_target"])
print("Best parameters found: ", grid_search_model.best_params_)

best_xgb_model = grid_search_model.best_estimator_  #set: best_xgb_model with best_estimator. 

y_pred_test = best_xgb_model.predict(test[predictors])
precision_score(test["BTC_target"], y_pred_test)
# 0.4846153846153846, 0.5833333333333334

[15:56:06] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[15:56:09] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[15:56:15] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[15:56:25] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[15:56:28] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "m

0.6333333333333333

Thử nghiệm kiểu training backtest training

In [112]:

# Đóng gói backtest function: 
def predict(train, test, predictors, model):
    model.fit(train[predictors], train["BTC_target"])
    preds = model.predict(test[predictors])
    preds = pd.Series(preds, index=test.index, name="predictions")
    combined = pd.concat([test["BTC_target"], preds], axis=1)
    return combined

# Evaluating error with backtesting
def backtest(data, model, predictors, start=1095, step=150): # 150 - every 6 months
    """
    backtest: train-test Time Series with HỌC TĂNG CƯỜNG. 
    MỖI VÒNG LẶP SẼ CÓ THÊM 150 HÀNG mới tham gia vào tập huấn luyện. 
    """
    all_predictions = []

    for i in range(start, data.shape[0], step):
        train = data.iloc[0:i].copy()
        test = data.iloc[i:(i+step)].copy()
        predictions = predict(train, test, predictors, model)
        all_predictions.append(predictions)
    
    return pd.concat(all_predictions)

In [115]:
# XGB with backtest def

# XGBoost with simple parameters
from xgboost import XGBClassifier # XGBoost, LightGBM, AdaBoost, CatBoost-Boosting-Ensemble Methods.  
xgb_model = XGBClassifier(random_state=1, learning_rate=.1, n_estimators=50)
predictions = backtest(data, xgb_model, predictors)
precision_score(predictions["BTC_target"], predictions["predictions"]) # tính độ chính xác của model sau các lần backtesting
# precision_score(test["BTC_target"], y_pred_test)

[16:02:51] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[16:02:51] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[16:02:51] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[16:02:52] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.

[16:02:52] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "m

0.5006765899864682

# 4 - SUM

Điều chỉnh siêu tham số BẰNG RANDOMIZED SEARCH ta thu được bộ parameters tối ưu nhất là: Best parameters found:  {'colsample_bytree': 0.5, 'gamma': 2.9191531825142496, 'learning_rate': 0.3, 'max_depth': 3, 'min_child_samples': 20, 'n_estimators': 100, 'random_state': 1, 'subsample': 0.7}. Với độ chính xác: 0.6333333333333333

Thử nghiệm với backtest trong XGB chưa thấy được nhiều hiệu quả.

Note: Có thể chưa phải bộ tham số tốt nhất vì ta dùng Randomized search nên có thể bỏ lỡ 1 số bộ tham số.  

In [117]:
# SIMPLE XGBClassifier Model when hyperparameters have not been adjusted 
from xgboost import XGBClassifier

xgb_model = XGBClassifier(random_state=1, colsample_bytree=0.5, gamma=2.9191531825142496, learning_rate= 0.3, max_depth= 3, min_child_samples=20, n_estimators=100, subsample = 0.7)
xgb_model.fit(train[predictors], train["BTC_target"])

y_pred_test = xgb_model.predict(test[predictors])
precision_score(test["BTC_target"], y_pred_test)



[16:05:06] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-07593ffd91cd9da33-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "min_child_samples" } are not used.



0.6333333333333333